### ISSUE TITLE
[4.0.0-beta1] Kodi not working

### ISSUE THREAD

**Issue by noj86** [recalbox/recalbox-os#726]
Hello All,

It's a fresh install on a raspberry pi 2, when you launch Kodi 15.2, it gives you the splash screen (Isengard stuff) than the screen flashes and goes back to emulation station. If you try with recalbox in french, you might get a glimpse of kodi ui before it does the same thing.

Cheers
J.

**Comment by HolyFailz** [comment]
I had to disconnect my PS4 controller and use a keyboard to get Kodi to work

**Comment by nadenislamarre** [comment]
can you confirm that you can reproduce the bug each time when the ps4 is plugged ?
can you connect via ssh to the recalbox and run /recalbox/scripts/recalbox-support.sh
then, send me the file generated in /recalbox/share/saves please ?

**Comment by HolyFailz** [comment]
[recalbox-support-20160226212245.tar.gz](https://github.com/recalbox/recalbox-os/files/148853/recalbox-support-20160226212245.tar.gz)

That was my first time SSHing so let me know if that was done correctly. If I go into Kodi and disable joystick and gamepad support I can keep my PS4 controller plugged in but obviously the controller won't function. If I turn gamepad support back on while the controller is connected it takes me back to the main screen of my games.

**Comment by nadenislamarre** [comment]
hum, it doesn't contain the information i want.
can you reconnect, type : 
cat ~/.kodi/userdata/keymaps/recalbox.xml
and give me the output please.

**Comment by HolyFailz** [comment]
[Kodi.txt](https://github.com/recalbox/recalbox-os/files/149004/Kodi.txt)

**Comment by johnlenfr** [comment]
Same problem for me. Kodi doesn't launch and stay on desktop kodi theme but without nothing except the cursor of the mouse on top left hand corner.

I'm using my samsung TV as sceen.

**Comment by rockaddicted** [comment]
I practiced tests on irc with an user on recalbox 4.0.0, and using a ps4 gamepad.
After removing joystick.* files from `/usr/share/kodi/system/keymaps/` he has not had anymore problems and he was able to control Kodi with his ps4 gamepad.

**Comment by johnlenfr** [comment]
Problem is that I don't have ps4 gamepad connected: only wireless mouse and keyboard.
Maybee I have to open another issue then?

I tested also with wired keybord without mouse but does'nt work either.

**Comment by noj86** [comment]
[recalbox-support-20160227180035.tar.gz](https://github.com/recalbox/recalbox-os/files/149439/recalbox-support-20160227180035.tar.gz)

This is the content of the support script ;)

I was indeed using a ps4 controller, i'll test if it works when it's not connected and come back to you too.

**Comment by noj86** [comment]
Alright, if I disconnect the ps4 controller it launches properly. I'll check what rockaddicted suggested too :)

**Comment by noj86** [comment]
Okay so here is what's happening : 

I've tried what rockaddicted suggested.

1/ ssh in the pi as root ( login : root password : recalboxroot)
2/ allow read/write on the system file using (at "/" location) : "mount -o -remount, rw /"
3/ moved the joystick.*.xml files avay from "/usr/share/kodi/system/keymaps" using "mv /usr/share/kodi/system/keymaps/joystick.*.xml /recalbox/share/kodi/."
4/ reboot, with the ps4 controller plugged in (usb) and kodi doesn't crash anymore

but

the mapping of the controller is completely broken :(

basically you get only this : 

X and O => works as expected (confirm for X, cancel for O)
L2 => Mute/unmute
Left Joystick up => increase volume
Left Joystick left => decrease volume
Right Joystick left => move the ui to the left

everything else, including the joysticks other direction and dpad does nothing.

**Comment by noj86** [comment]
Okay a bit more strange, without these keymaps file i can't return to emulation station basically, the shotdown option of kodi just hang kodi. The moment the keymaps files are back and I drive everything through the keyboard without any connected controller it works fine.

**Comment by noj86** [comment]
Just in case it's needed too, that the content of "cat ~/.kodi/userdata/keymaps/recalbox.xml"

<?xml version="1.0" ?>
<keymap>
        <global/>
        <FullscreenVideo/>
</keymap>

cheers

**Comment by nadenislamarre** [comment]
you mean that the file is only one line ?!?

**Comment by noj86** [comment]
yes

**Comment by nadenislamarre** [comment]
1) so, that somethings works is already a problem then.
2) then the file is only 1 line is a problem too

come on irc tonight please i'm susan33 under irc.

2016-02-27 19:08 GMT+01:00 noj86 <notifications@github.com>:

**Comment by noj86** [comment]
I'm on it already ;) just sent you an mp

**Comment by nadenislamarre** [comment]
tonight is 20h o clock, bbc time.

**Comment by noj86** [comment]
Okay after the help of nadeslamarre, we haven't fixed the problem but here is some more discovery:

The file that makes kodi crash is : /usr/share/kodi/system/keymaps/joystick.Sony.PLAYSTATION(R)3.Controller.xml
for now if you rename the extension it will prevent the crash of kodi.

We've also been through some tests of the mapping of the ps4 controller, so here it is : 

Up Arrow ==> ABS_HAT0Y -1
Down Arrow ==> ABS_HAT0Y 1
Left Arrow ==> ABS_HAT0X -1
Right Arrow ==> ABS_HAT0X 1
 
Up LS ==> ABS_Y 0
Down LS ==> ABS_Y 255
Left LS ==> ABS_X 0
Right LS ==> ABS_X 255
 
Up RS ==> ABS_RZ 0
Down RS ==> ABS_RZ 255
Left RS ==> ABS_Z 0
Right RS ==> ABS_Z 255

and the buttons are going from number 304 to 317

**Comment by palmipode** [comment]
hi same problem for me kodi stays on the wallpaper nothing more... 

Thanks for your help

**Comment by guizzzmo** [comment]
same problem kodi freeze on splash screen

**Comment by nadenislamarre** [comment]
and you have a ps4 too ?

2016-03-01 14:36 GMT+01:00 guizzzmo <notifications@github.com>:

**Comment by guizzzmo** [comment]
no i have just keyboard and 8 bitdo nes30 pro in bluetooth

**Comment by nadenislamarre** [comment]
and if you unplug the nes30 does it work ?

2016-03-01 16:33 GMT+01:00 guizzzmo <notifications@github.com>:

**Comment by guizzzmo** [comment]
i have don't try. i will try soon.

**Comment by palmipode** [comment]
no a ps3 sixaxis .

**Comment by rockaddicted** [comment]
On the forum, someone reported same problem with a xin-mo controller.

**Comment by nadenislamarre** [comment]
Current status :
- mainly occurs on the ps4 joystick but probably on some other pad (to be confirmed)
- it looks like that kodi associate some joystick configuration .xml to the ps4 for an unknown reason
- if you define a configuration for axis 6 in a configuration file but that your pad has only 4, kodi crash => it's why removing the joystick*.xml make kodi no more crash ; however this is not the solution (and it doesn't work, it just doesn't crash).
- a basic configuration with just hat in the recalbox.xml show that it's not taken into account
- it seems that only axis are affected (hat+axes while hat are axis in fact)
- in kodi, axis are seens, but when you push, it goes into no xml defined action associated code.
- i've analysed the kodi code. I've found thngs to check when a ps4 is plug, there are notably some strange parts which could not work on the pi while it's computation with binary operators when saving axis and hat configuration into hash map (sdl2 has a similar computation problem which made the shift key not working).
- required really a ps4 to investigate more. Possibly i'll get one for 2 days on 29 and 30 March to make analyses to continue investigation. (or if somebody on Bordeaux France has one to bring to me for some days ;-)

**Comment by nadenislamarre** [comment]
news :
- i have a ps4 gamepad for 3 days maximum, so i want feedback as soon as possible
- problem identified : not specific to ps4 gamepad but to events flooding (axis having very small movements permanently sent by the controller)
- the xml making crash at startup is an other issue, for the moment, you need to remove these xml to avoid kodi crashing at startup : /usr/share/kodi/system/keymaps/joystick*.xml
- it is a kodi bug. I can reproduce on kodi from 15 to the last 16 version, on the rpi and on the computer.

in 2 words, to simplify, the code includes things like :
if(something happend on axes) {
  do action
} else if (something happend on hats) {
 do action // never done if somethings always happend on axes, even if the move is almost 0
}
this is not as triavial, but basically the patch looks like removing the "else".

buttons are threated just before axis in a similar ifelse explaining why only buttons were working.
so i want feedback while continuing to fix.

new status :
- buttons works (were already working)
- hat works ! (you can then use kodi but without the 2 other axes)
- axes works strangly, not always (i have to handle the flooding inside axes events)
- xml have to be remove so that configuration reading doesn't crash

to test :
ssh on the recalbox,
mount -o remount,rw /
cd /usr/lib/kodi
rm kodi.bin
wget http://nadenislamarre.free.fr/ps4/kodi.bin.gz
gunzip kodi.bin.gz
rm  /usr/share/kodi/system/keymaps/joystick*.xml
mount -o remount,ro /
# start kodi

**Comment by nadenislamarre** [comment]
But 2 fixed
kodi allows triggers on joysticks, but they have to be configured
The fix includes an auto configuration based on the fact that when detecting pads, no button is pressed.

But 3 fixed :
we have effectively to remove the xml while kodi joystick information is based on the joystick name.
Unfornatly, some PS3 and PS4 controllers have the same joystick name something like "SONY PLAYSTATION ENTERTAINMENT SOMETHING", so kodi tries to apply the configuration of the PS3 pad to the PS4. While the axis are differents, it tries to configure an axe that doesnt exist (axis[7] when you've only 6 axes for example) making kodi crash.

I'll commit soon these 3 joystick kodi fixes.

**Comment by nadenislamarre** [comment]
Fixed for the next build.

**Comment by digitalLumberjack** [comment]
Anyone installed the unstable and tested ?

**Comment by digitalLumberjack** [comment]
(just set the updates.type=unstable in recalbox.conf)

**Comment by nadenislamarre** [comment]
to be closed is nobody does feedback.
Works for me now.

**Comment by digitalLumberjack** [comment]
@recalbox/recalbox-testers can you test that on 4.0.0 latest unstable ?

**Comment by nosfay** [comment]
I have no controls when i launch Kodi. My controller is a PS3 Shanwan. I have to exit via SSH.
I dont know if it's related to your fix because im not a Kodi user.

**Comment by nadenislamarre** [comment]
which recalbox version ?
4.0 beta2 fixes (hum, or beta3 will fixes, i don't remember)

**Comment by nosfay** [comment]
Last unstable

**Comment by benutzer193** [comment]
Working perfectly with rpi2+Logitech Rumblepad 2/Logitech DualAction

**Comment by Didlith** [comment]
Idem on rpi3 kodi working with xbox 360 wireless controleur

**Comment by davidkatana** [comment]
hello 

i'm not at home i will test this week end on the rasp 2 and 3 (i will got the new one this friday ! ) 

best regards 

david K 


De: "digitalLumberjack" <notifications@github.com> 
À: "recalbox/recalbox-os" <recalbox-os@noreply.github.com> 
Cc: "David KATANA" <d.katana@mancheopenschool.fr> 
Envoyé: Mardi 12 Avril 2016 13:16:32 
Objet: Re: [recalbox/recalbox-os] [4.0.0-beta1] Kodi not working (#726) 



@recalbox/recalbox-testers can you test that on 4.0.0 latest unstable ?

**Comment by nadenislamarre** [comment]
nosfay : my guess is that you've to wait the next beta (beta 3),
it should be out very soon.

**Comment by cr4nk89** [comment]
I updated the unstable but still not working. It stucks on the kodi homescreen. still schowing "pulse-eight cec adapter.." in the lower right corner.

**Comment by nadenislamarre** [comment]
it is supposed to work now.
can you do :
1) quit kodi (go to es)
2) ssh and mv ~/.kodi to ~/.kodi_backup
3) reboot to reinitialize the .kodi for recalbox
4) start kodi
and tell if that works. Perhaps your .kodi is broken now.

**Comment by cr4nk89** [comment]
By doing this it doesn't even open kodi anymore.

**Comment by davidkatana** [comment]
Hi all 
sorry i can't test your soft because my rpi crashed this week end.. 
i'm waiting the rpi3 but i haven't received yet 
i will come back 
Best regards 
David 


De: "cr4nk89" <notifications@github.com> 
À: "recalbox/recalbox-os" <recalbox-os@noreply.github.com> 
Cc: "David KATANA" <d.katana@mancheopenschool.fr>, "Team mention" <team_mention@noreply.github.com> 
Envoyé: Mardi 19 Avril 2016 22:07:05 
Objet: Re: [recalbox/recalbox-os] [4.0.0-beta1] Kodi not working (#726) 



By doing this it doesn't even open kodi anymore.

**Comment by digitalLumberjack** [comment]
Hi, ps3 controller doesn't works anymore in kodi.

**Comment by digitalLumberjack** [comment]
In fact it works when i reconfigure the pad :)

**Comment by digitalLumberjack** [comment]
i close this one
